File: Pattern Searching/Knuth_Morris_Pratt_Algorithm/KMP.java
/* Given a text txt[0..n-1] and a pattern pat[0..m-1], we need to find all occurrences of pat[] in txt[] */
/* Assume: n>m */
class KMP
{
    void KMPSearch(String pat, String txt)
    {
        int M = pat.length();
        int N = txt.length();
        int lps[] = new int[M]; // longest_prefix_suffix array
        int j = 0; // index for pat[]

        computeLPSArray(pat, M, lps); //calculate lps[] array

        int i = 0; //index for txt[]
        while(i < N)
        {
            if (pat.charAt(j) == txt.charAt(i))
            {
                j++;
                i++;
            }

            if (j == M)
            {
                System.out.println("Found pattern at index"+(i-j));
                j = lps[j - 1];
            }
            else if (i < N && pat.charAt(j) != txt.charAt(i))
            {
                if (j != 0)
                    j = lps[j - 1];
                else
                    i = i + 1;
            }
        }
    }

    void computeLPSArray(String pat, int M, int lps[]) // length of the previous longest prefix suffix
    {
        int len = 0;
        int i = 1;
        lps[0] = 0; //lps[0] is always 0

        while(i < M) //calculate lps[i] for i=1 to M-1
        {
            if (pat.charAt(i) == pat.charAt(len))
            {
                len++;
                lps[i] = len;
                i++;
            }
            else
            {
                if (len != 0)
                    len = lps[len - 1];
                else
                {
                    lps[i] = len;
                    i++;
                }
            }
        }
    }

    //Driver program to test above function
    public static void main(String args[])
    {
        String txt = "namanchamanbomanamansanam";
        String pat = "aman";
        new KMP().KMPSearch(pat, txt);
    }
}

/* Output

Pattern found at 1
Pattern found at 7
Pattern found at 16
*/

File: Algo_Ds_Notes/Knuth_Morris_Pratt_Algorithm/KMP.cs
using System;

namespace KMP_matching
{
    class Program
    {
        public static void calculateLps(string pattern, int[] lps)
        {
            int length = 0; // Length of the previous longest prefix suffix
            int i;

            lps[0] = 0;
            i = 1;

            // The loop calculates lps[i] for i = 1 to size - 1
            while(i < pattern.Length)
            {
                if(pattern[i] == pattern[length])
                {
                    length++;
                    lps[i] = length;
                    i++;
                }
                else // If (pattern[i] != pattern[len])
                {
                    if(length != 0)
                        length = lps[length - 1];
                    else
                    {
                        lps[i] = 0;
                        i++;
                    }
                }
            }
        }

        public static void KMPSearch(string pattern, string text)
        {
            int sizePattern = pattern.Length;
            int sizeText = text.Length;

            int[] lps = new int[sizePattern];
            int j  = 0;

            calculateLps(pattern, lps);

            int i = 0;

            while(i < sizeText)
            {
                if (pattern[j] == text[i])
                {
                    j++;
                    i++;
                }

                if(j == sizePattern)
                {
                    Console.WriteLine("Pattern found at " + Convert.ToString(i - j + 1));
                    j = lps[j - 1];
                }
                else if (i < sizeText && pattern[j] != text[i])
                {
                    if(j != 0)
                        j = lps[j - 1];
                    else
                        i = i + 1;
                }
            }
        }

        public static void Main(String[] args)
        {
            string text = "namanchamanbomanamansanam";
            string pattern = "aman";
            KMPSearch(pattern, text);
            Console.ReadLine();
        }
    }
}


/* Output

Pattern found at 2
Pattern found at 8
Pattern found at 17

*/
